torchvision/models/mnasnet.py

```diff
@@ -0,0 +1,166 @@
+        assert stride in [1, 2]
+        assert kernel_size in [3, 5]
+        mid_ch = in_ch * expansion_factor
+        self.apply_residual = (in_ch == out_ch and stride == 1)
+        self.layers = nn.Sequential(
+            # Pointwise
+            nn.Conv2d(in_ch, mid_ch, 1, bias=False),
+            nn.BatchNorm2d(mid_ch, momentum=bn_momentum),
+            nn.ReLU(inplace=True),
+            # Depthwise
+            nn.Conv2d(mid_ch, mid_ch, kernel_size, padding=kernel_size // 2,
+                      stride=stride, groups=mid_ch, bias=False),
+            nn.BatchNorm2d(mid_ch, momentum=bn_momentum),
+            nn.ReLU(inplace=True),
+            # Linear pointwise. Note that there's no activation.
+            nn.Conv2d(mid_ch, out_ch, 1, bias=False),
+            nn.BatchNorm2d(out_ch, momentum=bn_momentum))
+
+    def forward(self, input):
+        if self.apply_residual:
+            return self.layers.forward(input) + input
+        else:
+            return self.layers.forward(input)
+
+
+def _stack(in_ch, out_ch, kernel_size, stride, exp_factor, repeats,
+           bn_momentum):
+    """ Creates a stack of inverted residuals as seen in e.g. MobileNetV2 or
+    MNASNet. """
+    assert repeats >= 1
+    # First one has no skip, because feature map size changes.
+    first = _InvertedResidual(in_ch, out_ch, kernel_size, stride, exp_factor,
+                              bn_momentum=bn_momentum)
+    remaining = []
+    for _ in range(1, repeats):
+        remaining.append(
+            _InvertedResidual(out_ch, out_ch, kernel_size, 1, exp_factor,
+                              bn_momentum=bn_momentum))
+    return nn.Sequential(first, *remaining)
+
+
+def _round_to_multiple_of(val, divisor, round_up_bias=0.9):
+    """ Asymmetric rounding to make `val` divisible by `divisor`. With default
+    bias, will round up, unless the number is no more than 10% greater than the
+    smaller divisible value, i.e. (83, 8) -> 80, but (84, 8) -> 88. """
+    assert 0.0 < round_up_bias < 1.0
+    new_val = max(divisor, int(val + divisor / 2) // divisor * divisor)
+    return new_val if new_val >= round_up_bias * val else new_val + divisor
+
+
+def _scale_depths(depths, alpha):
+    """ Scales tensor depths as in reference MobileNet code, prefers rouding up
+    rather than down. """
+    return [_round_to_multiple_of(depth * alpha, 8) for depth in depths]
+
+
+class MNASNet(torch.nn.Module):
+    """ MNASNet, as described in https://arxiv.org/pdf/1807.11626.pdf.
+    >>> model = MNASNet(1000, 1.0)
+    >>> x = torch.rand(1, 3, 224, 224)
+    >>> y = model.forward(x)
+    >>> y.dim()
+    1
+    >>> y.nelement()
+    1000
+    """
+
+    def __init__(self, num_classes, alpha, dropout=0.2):
+        super(MNASNet, self).__init__()
+        self.alpha = alpha
+        self.num_classes = num_classes
+        self.dropout = dropout
+        depths = _scale_depths([24, 40, 80, 96, 192, 320], alpha)
+        layers = [
```

Comment:
The more I look at this, the more I have the impression it looks a lot like MobileNet V2 https://github.com/pytorch/vision/blob/74b6a750c7b748d03a3c80f29ebf17ca98dcab8b/torchvision/models/mobilenet.py#L49-L74

Would there be some benefits of making both implementations share a bit more building blocks?

Reply:
Forget about this comment for now, let's leave each file in `models` to be self-contained

Reply:
@fmassa same building blocks as MobileNetV2, FBNet, and ChamNet. I just realized this after doing my own impl of MnasNet yesterday to have the SE variant ('A1' by Google's naming). Then I noticed the comparisons in the FBNet paper, and then the FBNet backbone def in maskrcnn-benchmark (https://github.com/facebookresearch/maskrcnn-benchmark/blob/464b1af167d3dcb513964e62389825ff507609c3/maskrcnn_benchmark/modeling/backbone/fbnet_builder.py)

With maybe a few small tweaks all of the MNasNet architectures can be defined with the above building code, as can all of the other mentioned networks. It'd be nice to have them all together here with the same generic builder :)

Reply:
Speaking of SE, that would not be difficult to add here as a mode, though I never used it myself in practice.